core/src/main/java/org/apache/iceberg/metrics/CounterResultParser.java

```diff
@@ -0,0 +1,105 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import org.apache.iceberg.metrics.MetricsContext.Unit;
+import org.apache.iceberg.metrics.ScanReport.CounterResult;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+
+class CounterResultParser {
+
+  private static final String MISSING_FIELD_ERROR_MSG =
+      "Cannot parse counter from '%s': Missing field '%s'";
+
+  private static final String NAME = "name";
+  private static final String UNIT = "unit";
+  private static final String VALUE = "value";
+
+  private CounterResultParser() {}
+
+  static String toJson(CounterResult counter) {
+    return toJson(counter, false);
+  }
+
+  static String toJson(CounterResult counter, boolean pretty) {
+    return JsonUtil.generate(gen -> toJson(counter, gen), pretty);
+  }
+
+  static void toJson(CounterResult counter, JsonGenerator gen) throws IOException {
+    toJson(counter, gen, true);
+  }
+
+  static void toJson(CounterResult counter, JsonGenerator gen, boolean includeName)
+      throws IOException {
+    Preconditions.checkArgument(null != counter, "Invalid counter: null");
+
+    gen.writeStartObject();
+    // ScanMetricsResultParser mainly uses this with includeName=false, since it includes the name
+    // in a parent structure
+    if (includeName) {
+      gen.writeStringField(NAME, counter.name());
+    }
+    gen.writeStringField(UNIT, counter.unit().displayName());
+    gen.writeNumberField(VALUE, counter.value());
+    gen.writeEndObject();
+  }
+
+  static CounterResult fromJson(String json) {
+    return JsonUtil.parse(json, CounterResultParser::fromJson);
+  }
+
+  static CounterResult fromJson(JsonNode json) {
+    Preconditions.checkArgument(null != json, "Cannot parse counter from null object");
+    Preconditions.checkArgument(json.isObject(), "Cannot parse counter from non-object: %s", json);
+
+    String name = JsonUtil.getString(NAME, json);
+    String unit = JsonUtil.getString(UNIT, json);
+    long value = JsonUtil.getLong(VALUE, json);
+    return new CounterResult(name, Unit.fromDisplayName(unit), value);
+  }
+
+  /**
+   * This is mainly used from {@link ScanMetricsResultParser} where the counter name is already part
+   * of the parent {@link JsonNode}, so we omit checking and reading the counter name here.
```

Comment:
What was the purpose of adding a name to the counter? The handling here makes me think that we're not modeling correctly by having name in two places.

Reply:
Metrics having their own name is to align with other metric frameworks that associate a name with the metric independent of the internal variable name.  I feel the name should be used for the json representation and reporting.

Reply:
I agree with Dan here. The metric name is because other metric frameworks track it and I also think this should be part of serde+reporting